docs/logging-design.md

```diff
@@ -0,0 +1,51 @@
+# Design of the way AsyncHTTPClient logs
+
+<details>
+  <summary>Unless you explicitly pass AsyncHTTPClient a `Logger` instance, nothing is ever logged.</summary>
+  AsyncHTTPClient is useful in many places where you wouldn't want to log, for example a command line HTTP client. Also, we do not want to change its default behaviour in a minor release.
+</details>
+<details>
+  <summary>Nothing is logged at level `info` or higher, unless something is really wrong that cannot be communicated through the API.</summary>
+  Fundamentally, AsyncHTTPClient performs a simple task, it makes a HTTP request and communicates the outcome back via its API. In normal usage, we would not expect people to want AsyncHTTPClient to log. In certain scenarios, for example when debugging why a request takes longer than expected it may however be useful to get information about AsyncHTTPClient's connection pool. That is when enabling logging may become useful.
+</details>
+<details>
+  <summary>Each request will get a globally unique request ID (`ahc-request-id`) that will be attached (as metadata) to each log message relevant to a request.</summary>
+  When many concurrent requests are active, it can be challenging to figure out which log message is associated with which request. To facilitate this task, AsyncHTTPClient will add a metadata field `ahc-request-id` to each log message so you can first find the request ID that is causing issues and then filter only messages with that ID.
+</details>
+<details>
+  <summary>Your `Logger` metadata is preserved.</summary>
+  AsyncHTTPClient accepts a `Logger` on every request method. This means that all the metadata you have attached, will be present on log messages issued by AsyncHTTPClient.
+
+ For example, if you attach `["my-system-req-uuid": "84B453E0-0DFD-4B4B-BF22-3434812C9015"]` and then do two requests using AsyncHTTPClient, both of those requests will carry `"my-system-req-uuid` as well as AsyncHTTPClient's `ahc-request-id`. This allows you to filter all HTTP request made from one of your system's requests whilst still disambiguating the HTTP requests (they will have different `ahc-request-id`s.
```

Comment:
You may want to be specific how "merge" works, or leave it up to our context work to define it (also fine) if a key is overlapping, would both be stored, would the "latter" one win etc.

Reply:
@ktoso Not sure I fully understand. Right now, we don't do any merging in AHC. We just add some `ahc-*` keys.

Is what you're referring to the following situation:

```swift
// user code (outside AHC)
var logger = ...
logger[metadataKey: "connection"] = "a very good connection"

// in AHC
logger.debug("established connection", metadata: ["connection": "\(ahcConnection)"])
```

? If yes, then it's up to the `LogHandler` (the logging backend) to merge the on-Logger metadata with the ad-hoc metadata from a message.

Reply:
Hey right that's what I had in mind, though you're right that it's not AHC that has to care about it 👍 Ignore that comment here, it concerns log handler and tracers :)